cni/pkg/iptables/iptables.go

```diff
@@ -0,0 +1,495 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iptables
+
+import (
+	"errors"
+	"fmt"
+	"net/netip"
+	"strings"
+
+	"istio.io/istio/cni/pkg/nodeagent/constants"
+	istiolog "istio.io/istio/pkg/log"
+	"istio.io/istio/tools/istio-iptables/pkg/builder"
+	iptablesconfig "istio.io/istio/tools/istio-iptables/pkg/config"
+	iptablesconstants "istio.io/istio/tools/istio-iptables/pkg/constants"
+	dep "istio.io/istio/tools/istio-iptables/pkg/dependencies"
+	iptableslog "istio.io/istio/tools/istio-iptables/pkg/log"
+)
+
+const (
+	// INPOD marks/masks
+	InpodTProxyMark      = 0x111
+	InpodTProxyMask      = 0xfff
+	InpodMark            = 1337 // this needs to match the inpod config mark in ztunnel.
+	InpodMask            = 0xfff
+	InpodRestoreMask     = 0xffffffff
+	ChainInpodOutput     = "ISTIO_OUTPUT"
+	ChainInpodPrerouting = "ISTIO_PRERT"
+	ChainHostPostrouting = "ISTIO_POSTRT"
+	RouteTableInbound    = 100
+)
+
+var log = istiolog.RegisterScope("iptables", "iptables helper")
+
+type Config struct {
+	RestoreFormat     bool `json:"RESTORE_FORMAT"`
+	TraceLogging      bool `json:"IPTABLES_TRACE_LOGGING"`
+	EnableInboundIPv6 bool `json:"ENABLE_INBOUND_IPV6"`
+	RedirectDNS       bool `json:"REDIRECT_DNS"`
+}
+
+type IptablesConfigurator struct {
+	ext    dep.Dependencies
+	nlDeps NetlinkDependencies
+	cfg    *Config
+}
+
+func ipbuildConfig(c *Config) *iptablesconfig.Config {
+	return &iptablesconfig.Config{
+		RestoreFormat:     c.RestoreFormat,
+		TraceLogging:      c.TraceLogging,
+		EnableInboundIPv6: c.EnableInboundIPv6,
+		RedirectDNS:       c.RedirectDNS,
+	}
+}
+
+func NewIptablesConfigurator(cfg *Config, ext dep.Dependencies, nlDeps NetlinkDependencies) *IptablesConfigurator {
+	if cfg == nil {
+		cfg = &Config{
+			RestoreFormat: true,
+		}
+	}
+	return &IptablesConfigurator{
+		ext:    ext,
+		nlDeps: nlDeps,
+		cfg:    cfg,
+	}
+}
+
+func (cfg *IptablesConfigurator) DeleteInpodRules() error {
+	var inpodErrs []error
+
+	log.Debug("Deleting iptables rules")
+
+	inpodErrs = append(inpodErrs, cfg.executeDeleteCommands(), cfg.delInpodMarkIPRule(), cfg.delLoopbackRoute())
+	return errors.Join(inpodErrs...)
+}
+
+func (cfg *IptablesConfigurator) executeDeleteCommands() error {
+	deleteCmds := [][]string{
+		{"-t", iptablesconstants.MANGLE, "-D", iptablesconstants.PREROUTING, "-j", ChainInpodPrerouting},
+		{"-t", iptablesconstants.MANGLE, "-D", iptablesconstants.OUTPUT, "-j", ChainInpodOutput},
+		{"-t", iptablesconstants.NAT, "-D", iptablesconstants.OUTPUT, "-j", ChainInpodOutput},
+	}
+
+	// these sometimes fail due to "Device or resource busy"
```

Comment:
Deleting the line that jumps into our secondary chains (L432) is the one that we don't want to fail (unless it fails because it's not there, in which case, we are good)

Once that is gone, our created secondary chains are effectively dead/dangling and will never be processed, so if we cannot delete them on shutdown of the node agent, we know the node iptables rules will not be compromised.